### ISSUE TITLE
sclang: mouse position is complicated to get

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#2236]
While it is a standard way to control UGens by `MouseX` and `MouseY`, it is really quite complicated to get the same values from sclang. The easiedst seems to be:

`SkipJack({ defer { ~cursor = QtGUI.cursorPosition } }, 0.1);`

and then refer to `~cursor`. But this is far too much hassle for such a basic thing. I think this should be hidden from the user.

**Comment by telephon** [comment]
E.g.

```
MousePosition.x(0, 10)
```


```

MousePosition {

	classvar <>updatePeriod = 0.1, <position, task, screenFactor;

	*init {
		task ?? {
			position = Point(0, 0);
			screenFactor = Window.screenBounds.extent;
			task = SkipJack({
				screenFactor = Window.screenBounds.extent;
				position = QtGUI.cursorPosition;
			}, { updatePeriod }, false, "MousePosition", AppClock)
		}
	}

	*stop {
		task.stop
	}

	*x { |minval = 0, maxval = 1, warp = 0|
		this.init;
		^this.position.x.linlin(0, screenFactor.x, minval, maxval)
	}

	*y { |minval = 0, maxval = 1, warp = 0|
		this.init;
		^this.position.y.linlin(0, screenFactor.y, minval, maxval)
	}
}
```

**Comment by scztt** [comment]
Yeah, that definitely looks worthwhile. A few comments:
- Explicit MousePosition:start and MousePosition:stop, with no implicit init call? (I don't like the idea of starting a routine in the background just by accessing a property getter).
- The min/max/warp makes it impossible get a plain screen-coordinates mouse position. What if the default (e.g. minval.isNil && maxval.isNil) is no warp?
- The min screen coordinates are not necessarily zero. MacOS, at least, counts from the top-left corner of the main display - so, if your main display is the rightmost of two, then the coordinates of the leftmost screen are negative.
- Accessing via getters isn't functionally much different from calling (the more verbose) ```Window.screenBounds.extent.x.linlin(...)```. If it also notified listeners, then you wouldn't need to create an additional Routine / SkipJack to simply poll a value that's already polled:

```
SkipJack({
   newPos = Window.screenBounds.extent;
   screenFactor = Window.screenBounds.extent;
   if (position != newPos) {
      position = newPos;
      this.changed(\value, position);
   };
})
```

This would allow e.g.:
```MousePosition.addDependent({ |obj, what, pos| ~numberBoxX.value = pos.x; ~numberBoxY.value = pos.y });```

- updatePeriod shouldn't be in a classvar - otherwise changing it could break the usage of other clients (e.g. a Quark that's using it). Actually, speaking of the auto-start behavior, automatically starting it feels much more acceptable if it's done for instances rather than implicitly in the class. It would also allow you to specify min/max as part of the constructor, e.g.
```
MousePosition { *new { |updatePeriod, spec, autoStart=true| ^super.newCopyArgs(updatePeriod, spec.asSpec).start(autoStart); } }
```

**Comment by telephon** [comment]
You are right, and I agree in particular about not starting a hidden task that can't be stopped. Instances would be better, but also more tasks. All that task stuff is really only there because  we need to call the UI thread. But this may not be necessary. I've had a look at how the `MouseX` is implemented, but didn't get far with tracing down those system calls.

**Comment by scztt** [comment]
I checked a bit and the QT mouse position calls DEFINITELY have to be called from the UI thread. We could make it a blocking call from the non-UI thread, but the non-UI thread is the one used for high-precision music events, so introducing hidden locks there is ultimately a bad idea (especially for operations that seem like they would be trivial e.g. mouse position).

This actually turns out to be a good test-case for the Connection quark I've been working on - it's something conceptually simple to hook up, but requires a ton of boilerplate scaffolding to do with the default classlib. Because Point's get turned into arrays when passed as a synth arg, you can pass the point directly to the synth, so it's very simple:
```
~synth = { |xy=#[100,100]| LFSaw.ar(xy) }.play;
~connection = PeriodicUpdater(QtGUI, \cursorPosition)
    .signal(\value).connectTo(~synth.argSlot(\xy));
```

Scaling can happen in the PeriodicUpdater:
```PeriodicUpdater({ QtGUI.cursorPosition / 300 + (100@300) })```
or after the fact ala:
```
~connection.transform({
    |o, w, value|
   [o, w, value / 300 + (100@300)]
});
```

If Point didn't conveniently convert to a synth arg (I'm actually not totally sure where that's happening), it would be useful to have something that passed along changes to some property of a changed value, rather than the value itself, e.g.
```
~position = PeriodicUpdater(QtGUI, \cursorPosition);
~position.signal(\value).propertyFilter(\x).connectTo(~synth.argSlot(\x));
~position.signal(\value).propertyFilter(\y).connectTo(~synth.argSlot(\y));
```

or better:
```
~position.signal(\value)
    .propertyFilters(\x, \y)
    .connectEach(~synth.argSlots(\x, \y));
```

**Comment by telephon** [comment]
This looks like an interesting quark, I'll definitely have a closer look. In the current case, there is no need to send the value to a synth really, because we have a UGen that does it. But certainly for other GUI parameters it is a good idea.

Is there a simple scheme that could be derived from this which would give us the cursor position in sclang "more directly"? I mean in a way, the ScIDE class could update a variable, but this may mean a constant (even if low) load.

**Comment by adcxyz** [comment]
Hi, I played with this idea a little, and here's a take on it:

```
/* TESTS: 
// use just the Mouse class (no polling needed):
Mouse.pos
Mouse.x
Mouse.y

// start polling - more efficient with instances
Mouse.start;

// use instances with independent actions:
// args are mouse, xval, yval, xraw, yraw
Mouse(\a, { |...args| args.postln });

// post just x
Mouse(\a, { |mouse, x| x.round(0.0001).postln });

// try to post x and y - but forgot y -> error
// -> posts "Mouse(\a).action failed."
Mouse(\a, { |mouse, x| [x, y].round(0.0001).postln });

// see the error
Mouse(\a).doAction;

// fix it
Mouse(\a, { |mouse, x, y| [x, y].round(0.0001).postln });

// second mouse instance, use spec for x
Mouse(\b, { |mouse, x| x.round(0.01).postln }, [20, 2000, \exp]);

Mouse(\a).free;
// b remains
Mouse(\b).free;

Mouse.free;
*/
//////////////// save as Mouse.sc ----//// 
Mouse {
	classvar <all, <task, <screenSize, pos, <prevPos;
	var <name, <>action, <>xspec, <>yspec;

	*initClass {
		all = ();
		screenSize = Window.screenBounds.extent;
	}

	*pos {
		if (task.isPlaying) { ^pos };
		^pos = QtGUI.cursorPosition;
	}

	*initTask {
		task ?? {

			task = SkipJack({
				pos = QtGUI.cursorPosition;
				if (pos != prevPos) {
					this.doActions;
				};
				prevPos = pos;
			}, 0.02, false, "Mouse", AppClock)
		}
	}

	*doActions {
		all.do { |mouse|
			try {
				mouse.doAction
			} {
				"%.action failed.\n".postf(mouse)
			}
		}
	}

	*free {
		task.stop;
		task = nil;
	}

	*start { this.initTask; task.play }
	*stop { task.stop }

	*x { ^this.pos.x }
	*y { ^this.pos.y }

	*xuni { ^(pos.x / screenSize.x) }
	*yuni { ^(pos.y / screenSize.y) }

	*new { |name, action, xspec, yspec|
		var res;
		this.initTask; // lazy init task here

		res = all[name];
		if (res.notNil) {
			action !? { res.action = action };
			xspec !? { res.xspec = xspec.asSpec };
			yspec !? { res.yspec = yspec.asSpec };
			^res
		};

		^super.newCopyArgs(name, action, xspec, yspec).init.prAdd;
	}
	init {
		xspec !? { xspec = xspec.asSpec };
		yspec !? { yspec = yspec.asSpec };
	}
	prAdd { all.put(this.name, this) }

	free { action = nil; all.removeAt(name) }

	xval {
		var xval = Mouse.xuni;
		xspec !? { xval = xspec.map(xval) };
		^xval
	}
	yval {
		var yval = Mouse.xuni;
		yspec !? { yval = yspec.map(yval) };
		^yval
	}

	doAction {
		action.value(this, this.xval, this.yval, Mouse.x, Mouse.y);
	}

	storeArgs { ^[name] }
	printOn { |str| this.storeOn(str) }
}
```

**Comment by avdrd** [comment]
I'm going to add here that from my experience, on Windows at least, the GUI (AppClock) thread in SC runs at such low priority that it doesn't do more than 10 FPS. I've written a tracker thiggie that moves the cursor of a SoundFileView as it plays, and 10 FPS is the best I can get at least when OSC callbacks are also used.

So a 10 FPS mouse update doesn't seem like an (additional) issue given that.

By the way, hitting the AppClock with insta-rescheduling like `AppClock.play(0)` doesn't use more than 30% of *one CPU core* on my box. So that's probably why it's so FPS-limited.
